Add sqlalchemy to the pipeline

CaselIT opened the pull request:
Fixes: #488

NOTE: this is currently broken when using the file from pypi because of https://github.com/sqlalchemy/sqlalchemy/issues/9106

Since this pipeline does not run on PR, you can use https://github.com/CaselIT/psycopg/actions/workflows/3rd-party-tests.yml to see it running

CaselIT commented on .github/workflows/3rd-party-tests.yml:
```diff
@@ -0,0 +1,89 @@
+      PSYCOPG_IMPL: ${{ matrix.impl }}
+      DEPS: ./psycopg[test] ./psycopg_pool pytest pytest-xdist
+
+    # could not translate host name "pg" to address: Temporary failure in name resolution
```
remove comment

CaselIT commented on .github/workflows/3rd-party-tests.yml:
```diff
@@ -0,0 +1,89 @@
+  push:
+    branches:
+      - "master"
+      - "sqlalchemy_pipeline"
```
this branch can be removed (or not)

dvarrazzo replied:
I don't see this branch in your repo. Should it be removed?

CaselIT replied:
it's definetly there, I'm using it to run the pipeline in my fork https://github.com/caselit/psycopg/tree/sqlalchemy_pipeline

My comment was just to mention that I can remove before the merge

CaselIT commented on .github/workflows/3rd-party-tests.yml:
```diff
@@ -0,0 +1,89 @@
+  cancel-in-progress: true
+
+jobs:
+  sqlalchemy:
```
probably need a custom name, since the default ones are not clear at all

dvarrazzo commented on .github/workflows/3rd-party-tests.yml:
```diff
@@ -0,0 +1,88 @@
+      fail-fast: false
+      matrix:
+        python-version:
+          - "3.11"
```
Only this one? Maybe the newest and the oldest supported (3.7 for pg3).

CaselIT replied:
Makes sense

dvarrazzo commented on .github/workflows/3rd-party-tests.yml:
```diff
@@ -0,0 +1,88 @@
+          - "3.11"
+        sqlalchemy:
+          - git+https://github.com/sqlalchemy/sqlalchemy.git#egg=SQLAlchemy
+          - sqlalchemy
```
This, at the moment, means 1.4.x and will become 2.0.x once the latter is out of beta, right? Maybe it's better to be more explicit?

CaselIT replied:
The pip command is using `--pre` so it's installing v2 already. but we can add `>=2`

dvarrazzo commented on .github/workflows/3rd-party-tests.yml:
```diff
@@ -0,0 +1,88 @@
+          pip install .
+
+      - name: Run sqlalchemy tests
+        run: bash -c "cd sa_home/$( ls sa_home ) &&
+          pytest -n 2 -q --dburi postgresql+psycopg://postgres:password@127.0.0.1/test --backend-only --dropfirst
+          "
```
Multiline would be more readable:

```suggestion
        run: |
          cd sa_home/$( ls sa_home )
          pytest -n 2 -q --backend-only --dropfirst \
             --dburi postgresql+psycopg://postgres:password@127.0.0.1/test

```

CaselIT replied:
I was experimenting, since GitHub doesn't show the output of the test until they are done for some reason.

I wanted to try another alternative to see if it works, if it doesn't I'll use what you suggested

dvarrazzo commented on .github/workflows/3rd-party-tests.yml:
```diff
@@ -0,0 +1,88 @@
+      - name: Setup PG
+        env:
+          PGPASSWORD: password
+        run: |
+          run_sql="psql -h 127.0.0.1 -d test -U postgres -c "
+          $run_sql 'SELECT version();'
+          $run_sql 'CREATE SCHEMA test_schema; CREATE SCHEMA test_schema_2;CREATE EXTENSION hstore;'
```
They can be run all together in the same command. We can use -e to echo the commands run. I also prefer to specify a connection string instead of the loose parameters.

```suggestion
        run: |
          psql -AXqte "host=127.0.0.1 dbname=test user=postgres" << HERE
          SELECT version();
          CREATE SCHEMA test_schema;
          CREATE SCHEMA test_schema_2;
          CREATE EXTENSION hstore;
          HERE
```

This would produce an output like:

```
SELECT version();
PostgreSQL 15.1 (Ubuntu 15.1-1.pgdg22.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0, 64-bit
CREATE SCHEMA test_schema;
CREATE SCHEMA test_schema_2;
CREATE EXTENSION hstore;
```

CaselIT replied:
thanks, that looks better!

dvarrazzo commented:
I seem to understand that there isn't a way to run this pipeline unless merged to master, or something like that? I have had this problem whenever I have added a new pipeline.

CaselIT commented:
> I seem to understand that there isn't a way to run this pipeline unless merged to master, or something like that? I have had this problem whenever I have added a new pipeline.

Yes, since otherwise anyone could create a pr adding a workflow that does whatever without the owner control.
You can look at the link I added above to see some runs

CaselIT commented:
I've updated.

On sqlalchemy side we were thinking about releasing an rc3. Once that's live this should pass all jobs (it's currently correctly failing on the version taken from pypi that's rc2, and passing on main)

CaselIT commented:
Sqlalchemy rc3 was released yesterday, so all jobs should pass now

dvarrazzo commented on .github/workflows/3rd-party-tests.yml:
```diff
@@ -0,0 +1,110 @@
+
+    steps:
+      - name: Select sqlalchemy pip target
+        run: |
+          case ${{ matrix.sqlalchemy }} in
+            git_main)
+              echo "PIP_SQLALCHEMY=git+https://github.com/sqlalchemy/sqlalchemy.git#egg=SQLAlchemy" >> $GITHUB_ENV
+              ;;
+            release)
+              echo "PIP_SQLALCHEMY=sqlalchemy>=2a" >> $GITHUB_ENV
+              ;;
+            *)
+              echo "Unknown value for sqlalchemy: ${{ matrix.sqlalchemy }}"
+              exit 1
+              ;;
+          esac
```
I would prefer to have these entries directly in the matrix and used later at `pip download`, without this intermediate step.

CaselIT replied:
I did this for the name of the job. I'm not sure if I can otherwise create something meaningful. Maybe GitHub supports json objects as variable values? I'll do somr tries

CaselIT replied:
I may have found a reasonable solution to have both a label and the variable defined in the matrix section

dvarrazzo commented on .github/workflows/3rd-party-tests.yml:
```diff
@@ -0,0 +1,110 @@
+
+    env:
+      PSYCOPG_IMPL: ${{ matrix.impl }}
+      DEPS: ./psycopg[test] ./psycopg_pool pytest pytest-xdist
```
I think we don't need the `[test]` of `psycopg` here, and maybe we don't need `psycopg_pool` either, unless you use it?

dvarrazzo commented on .github/workflows/3rd-party-tests.yml:
```diff
@@ -0,0 +1,110 @@
+
+    env:
+      PSYCOPG_IMPL: ${{ matrix.impl }}
+      DEPS: ./psycopg[test] ./psycopg_pool pytest pytest-xdist
```
Out of curiosity, why don't you have a `[test]` extra yourself, specifying `pytest-xdits`? You can then `pip install .[test]` in the `sa` dir to install library and test dependencies.

CaselIT replied:
I honestly did not look if sqlalchemy had one.

Maybe we can just use tox here? I'll take a look at what target there are

dvarrazzo replied:
It doesn't seem it has one, no.

TBH Tox has caused me so many problems that doing without it would be preferable for me.

CaselIT replied:
looking at the defined env there is one that could maybe be used, that's the github one, but running pytest should be easier

CaselIT commented on .github/workflows/3rd-party-tests.yml:
```diff
@@ -0,0 +1,110 @@
+        run: |
+          pip download --pre --no-deps --no-binary :all: $PIP_SQLALCHEMY
+          mkdir sa_home
+          if [[ ${{ matrix.sqlalchemy }} = "release" ]]
+          then
+            extract="tar -C sa_home -xzf"
+          else
+            extract="unzip -d sa_home -q"
+          fi
+          $extract $( ls SQLAlchemy* )
```
I use the sdist to also run the tests, so in any case it needs to be extracted

dvarrazzo replied:
> I use the sdist to also run the tests, so in any case it needs to be extracted

ah sure, that makes sense.

CaselIT commented:
sorry forgot to mention yesterday. I've updated

dvarrazzo commented:
Merged, thank you very much!

CaselIT commented:
I did not know that github automatically named the matrix jobs for you. It even does a better job that what I had before
![image](https://user-images.githubusercontent.com/16175304/216764187-6f2664b4-1a5a-4c37-a2c7-bb89590e562b.png)
